### PR TITLE
Admin emailer

### PR DESCRIPTION
**Pull request by mnussbaumer**
I added two basic tests.

And not really sure about the structure, the reason I call the .deliver_later on the mailer method itself is because there's really no penalty since bamboo spins a new process for it (so I imagine it will always be deliver_later that gets called, and because in case there were no admins the code in the controller would have to account for that before calling deliver_now, otherwise bamboo will raise an error for not having recipients.

In rails I would probably write an after_create callback for this but no sure how's the idiomatic way of doing that in elixir?

It's perhaps a bit unconventional. Take a look and let me know,
Micael

### CODE REVIEW

**Comment by odarriba** [comment]
Solves #3

**Comment by odarriba** [comment]
First of all, thanks a lot, seems very good code. I'm going to review it to propose a couple of changes, but the overall is very good!!

File: config/dev.exs [hunk 1]
```diff
@@ -55,5 +55,17 @@ config :elixir_jobs, ElixirJobsWeb.Guardian,
   issuer: "ElixirJobs",
   secret_key: "MY_D3V_K3Y"
 
+config :elixir_jobs, ElixirJobsWeb.Mailer,
```
Comment (odarriba) [inline config/dev.exs:58]: In development, I'd rather use the `Bamboo.LocalAdapter` adapter, which allows to integrate this: [EmailPreviewPlug](https://hexdocs.pm/bamboo/Bamboo.EmailPreviewPlug.html).

It makes the development much easier.

File: config/dev.exs [hunk 2]
```diff
@@ -55,5 +55,17 @@ config :elixir_jobs, ElixirJobsWeb.Guardian,
...
+  adapter: Bamboo.SMTPAdapter,
+  server: System.get_env("SMTP_SERVER"),
+  port: 587,
+  username: System.get_env("SMTP_USERNAME"),
+  password: System.get_env("SMTP_PASSWORD"),
+  tls: :if_available, # can be `:always` or `:never`
+  ssl: false, # can be `true`
+  retries: 1
+
+config :elixir_jobs, :home_url, "http://localhost:4000/"
```
Comment (odarriba) [inline config/dev.exs]: This is not really needed, we can generate URLs using `*_url` methods

File: config/prod.exs [hunk 3]
```diff
@@ -44,6 +44,18 @@ config :elixir_jobs, ElixirJobsWeb.Guardian,
...
+  adapter: Bamboo.SMTPAdapter,
+  server: System.get_env("SMTP_SERVER"),
+  port: 587,
+  username: System.get_env("SMTP_USERNAME"),
+  password: System.get_env("SMTP_PASSWORD"),
+  tls: :if_available, # can be `:always` or `:never`
+  ssl: false, # can be `true`
+  retries: 1
+
+config :elixir_jobs, :home_url, "https://elixirjobs.net/"
```
Comment (odarriba) [inline config/prod.exs]: Same with this one!

File: config/test.exs [hunk 4]
```diff
@@ -16,5 +16,12 @@ config :elixir_jobs, ElixirJobsWeb.Guardian,
   issuer: "ElixirJobs",
   secret_key: "MY_T3ST_K3Y"
 
+config :elixir_jobs, ElixirJobsWeb.Mailer,
+  adapter: Bamboo.TestAdapter
+
+config :bamboo, :refute_timeout, 3
+
+config :elixir_jobs, :home_url, "http://localhost:4000/"
```
Comment (odarriba) [inline config/test.exs]: ... and this!

File: lib/elixir_jobs/email.ex [hunk 5]
```diff
@@ -0,0 +1,31 @@
+defmodule ElixirJobsWeb.Email do
+  #import Bamboo.Email
+  use Bamboo.Phoenix, view: ElixirJobsWeb.EmailView
+
+  def notification_offer_created_html(offer) do
+    emails = ElixirJobs.Users.Admin |> ElixirJobs.Users.Queries.Admin.admin_emails |> ElixirJobs.Repo.all
```
Comment (odarriba) [inline lib/elixir_jobs/email.ex]: In the project we are using Phoenix 1.3.0 with Contexts, so the best way to do this is moving it to a function inside the `Users` context, which all the logic lives.

If you are okay with just getting the administrators and then get their e-mail addresses (so avoid extra code for each case):

```elixir
emails =
  ElixirJobs.Users.list_admins()
  |> Enum.map(&({&1.name, &1.email}))
```

File: lib/elixir_jobs/email.ex [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+  def notification_offer_created_text(emails, offer) do
+    put_basic_layouts(emails)
+    |> subject("A new job offer was posted")
+    |> render("offer_created.text", offer: offer)
+  end
+
+  defp put_basic_layouts(emails) do
+    new_email()
+    |> to(emails)
```
Comment (odarriba) [inline lib/elixir_jobs/email.ex]: Here I'd do one of this two things:

1. Use BCC for e-mails.
2. Sent an e-mail per admin.

Both are okay, and in both an admin doesn't see the rest's email addresses.

File: lib/elixir_jobs_web/controllers/offer_controller.ex [hunk 7]
```diff
@@ -34,7 +34,8 @@ defmodule ElixirJobsWeb.OfferController do
 
   def create(conn, %{"offer" => offer_params}) do
     case Offers.create_offer(offer_params) do
-      {:ok, _offer} ->
+      {:ok, offer} ->
+        send_creation_email(offer)
```
Comment (odarriba) [inline lib/elixir_jobs_web/controllers/offer_controller.ex]: I'd just call `ElixirJobsWeb.Email.notification_offer_created_html(offer)` directly here, as in the `send_creation_email` there is no extra logic.

File: test/elixir_jobs/users/users_test.exs [hunk 8]
```diff
@@ -26,7 +26,7 @@ defmodule ElixirJobs.UsersTest do
 
     test "get_admin!/1 returns the admin with given id" do
       admin = admin_fixture()
-      assert Users.get_admin!(admin.id).__struct__ == ElixirJobs.Users.Admin
+      assert Users.get_admin_by_id!(admin.id).__struct__ == ElixirJobs.Users.Admin
```
Comment (odarriba) [inline test/elixir_jobs/users/users_test.exs:29]: Good one!

File: lib/elixir_jobs/email.ex [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    |> from({"ElixirJobs", "elixirjobs@email.com"})
```
Comment (odarriba) [inline lib/elixir_jobs/email.ex]: I'd change this to some var on `config.exs`, and defaulting to `no-reply@elixirjobs.net`

**Review by odarriba — COMMENTED** [review]
Very good PR!

There are some changes to be made, but the overall is pretty good! I like it!

**Comment by mnussbaumer** [comment]
@odarriba only thing I'm having problems with is ** (UndefinedFunctionError) function ElixirJobsWeb.Bamboo.SentEmailViewerPlug.init/1 is undefined (module ElixirJobsWeb.Bamboo.SentEmailViewerPlug is not available)
I've tried with Bamboo.EmailPreviewPlug as well. 

Do you know why this is happening and if I need to change something else?

I've updated bamboo to 0.8, and the smtp adapter to 1.4.1

**Comment by odarriba** [comment]
Woops! I thought they'd released the 1.0.0

In the 0.8.0 the correct plug is [Bamboo.EmailPreviewPlug](https://hexdocs.pm/bamboo/Bamboo.EmailPreviewPlug.html).

I like your changes a lot, it's better now. The only missing thing is that in some point you call `ElixirJobsWeb.Email.notification_offer_created_html({offer, :default})` and that `:default` function signature is not defined anywhere (or I couldn't see it in the diff).

Amazing work.
Thanks!

**Comment by mnussbaumer** [comment]
@odarriba but I've tried with Bamboo.EmailPreviewPlug as well:

```
== Compilation error in file lib/elixir_jobs_web/router.ex ==
** (UndefinedFunctionError) function ElixirJobsWeb.Bamboo.EmailPreviewPlug.init/1 is undefined (module ElixirJobsWeb.Bamboo.EmailPreviewPlug is not available)
    ElixirJobsWeb.Bamboo.EmailPreviewPlug.init([])
    (phoenix) lib/phoenix/router/route.ex:98: Phoenix.Router.Route.build_dispatch/1
    (phoenix) lib/phoenix/router/route.ex:84: Phoenix.Router.Route.build_route_match/2
    (phoenix) lib/phoenix/router/route.ex:56: Phoenix.Router.Route.exprs/1
    (phoenix) lib/phoenix/router.ex:320: anonymous fn/1 in Phoenix.Router."MACRO-__before_compile__"/2
    (elixir) lib/enum.ex:1255: Enum."-map/2-lists^map/1-0-"/2
    (elixir) lib/enum.ex:1255: Enum."-map/2-lists^map/1-0-"/2
    (phoenix) expanding macro: Phoenix.Router.__before_compile__/1
    lib/elixir_jobs_web/router.ex:1: ElixirJobsWeb.Router (module)
    (elixir) lib/kernel/parallel_compiler.ex:121: anonymous fn/4 in Kernel.ParallelCompiler.spawn_compilers/1
```

```
scope "/", ElixirJobsWeb do
    pipe_through :browser # Use the default browser stack

    get "/", OfferController, :index
    get "/about", PageController, :about
    get "/rss", OfferController, :rss
    get "/page/:page", OfferController, :index, as: :offer_page
    get "/offers/new", OfferController, :new
    post "/offers/new", OfferController, :create
    post "/offers/preview", OfferController, :preview
    put "/offers/preview", OfferController, :preview
    get "/offers/:slug", OfferController, :show

    get "/login", AuthController, :new
    post "/login", AuthController, :create

    if Mix.env == :dev do
      forward "/sent_emails", Bamboo.EmailPreviewPlug
    end
  end
```

(I've added it to the non admin routes since it's only for dev environment it should be fine).
I've tried with adding `use Phoenix.Router` on the router as well.

Regarding bamboos version if I try to use the one from master in github it gives me this:

Dependencies have diverged:
* bamboo (https://github.com/thoughtbot/bamboo.git)
  the dependency bamboo in mix.exs is overriding a child dependency:

  > In mix.exs:
    {:bamboo, [env: :prod, git: "https://github.com/thoughtbot/bamboo.git", manager: :mix]}

  > In deps/bamboo_smtp/mix.exs:
    {:bamboo, "~> 0.8.0", [env: :prod, hex: "bamboo", repo: "hexpm", optional: false]}

  Ensure they match or specify one of the above in your deps and set "override: true"
** (Mix) Can't continue due to errors on dependencies

(I can't figure out where I specify and where do I set override:true)
Thanks

**Comment by odarriba** [comment]
The error of the plug seems to be caused because the block 
```
    if Mix.env == :dev do
      forward "/sent_emails", Bamboo.EmailPreviewPlug
    end
```

needs to be outside the `scope` one. That's because every module referenced inside the scope will be scoped inside the module of the scope (what a sentence, right?). In this case, `ElixirJobsWeb`).

**Comment by odarriba** [comment]
For the second error, I would recommend using the 0.8.0 because that's the stable one and the one in Github may contain bugs as is not released yet (only in beta I think). 

Changing it in `mix.exs` should be fine 😄

**Comment by mnussbaumer** [comment]
@odarriba - I meant changing the dep to follow the one in github resulted in that error, with 0.8 it works fine. Still with block or no block it doesn't seem to be able to find the Bamboo plug?

** (UndefinedFunctionError) function ElixirJobsWeb.Bamboo.EmailPreviewPlug.init/1 is undefined (module ElixirJobsWeb.Bamboo.EmailPreviewPlug is not available)

**Comment by mnussbaumer** [comment]
I mean it's trying to call init/1, but I'm not passing any arguments, so not sure what is going on there?

**Comment by odarriba** [comment]
That's the correct behavior: if you don't pass parameters to it, the call is made with an empty list.

The issue here is that it's trying to call `ElixirJobsWeb.Bamboo.EmailPreviewPlug.init/1` instead of `Bamboo.EmailPreviewPlug.init/1`.

With something like this it doesn't work?

```elixir
  scope "/", ElixirJobsWeb do
    pipe_through :browser # Use the default browser stack

    get "/", OfferController, :index
    get "/about", PageController, :about
    get "/rss", OfferController, :rss
    get "/page/:page", OfferController, :index, as: :offer_page
    get "/offers/new", OfferController, :new
    post "/offers/new", OfferController, :create
    post "/offers/preview", OfferController, :preview
    put "/offers/preview", OfferController, :preview
    get "/offers/:slug", OfferController, :show

    get "/login", AuthController, :new
    post "/login", AuthController, :create
  end

  if Mix.env == :dev do
    forward "/sent_emails", Bamboo.EmailPreviewPlug
  end
```

If not, can you please commit what you have so we can take a look? 🤔 
Come on! We are close to the merge!

**Comment by mnussbaumer** [comment]
@odarriba my bad I didn't understand what you meant by the do block sincerely lol. Thank you
let me just double check one more thing and I'll push the latest

**Comment by mnussbaumer** [comment]
@odarriba , so one last thing, I need to use the admin_offer_url/:edit since offer_url/:show won't find it because it's still unpublished (probably there should be a fallback when no slug is found in the public route?). 

But when I try to log in as an admin (If I'm not logged in), I always get a blank screen and this on iex
```
[error] Cowboy returned 400 and there are no headers in the connection.

This may happen if Cowboy is unable to parse the request headers,
for example, because there are too many headers or the header name
or value are too large (such as a large cookie).

You can customize those values when configuring your http/https
server. The configuration option and default values are shown below:

    protocol_options: [
      max_header_name_length: 64,
      max_header_value_length: 4096,
      max_headers: 100
    ]
```
If I visit the url again it works fine (but the url is lost). Although unrelated I can't seem to find why this is happening? I looked at ElixirJobsWeb.AuthController.create and it seems to be well setup?

**Comment by odarriba** [comment]
I think that's a separate issue... Maybe a environment bug (too large cookies stored in your browser for localhost or something)

**Comment by mnussbaumer** [comment]
Ah indeed. Can you check the PR and let me know?

**Comment by odarriba** [comment]
It's perfect! Thanks a lot for your work. I'll add a contributors list and your name will be there for sure!

Merging and closing #3 !

**Pull request merged**